### DISCUSSION TITLE
Allow passing type name as a function parameter without typeof.

### DISCUSSION THREAD

**Discussion by edward-a** [dotnet/csharplang#2804] [Language Ideas]
E.g.
````csharp
class A
{
    void Method(object x) {
    }

    void Run() {
        Method(A); // Instead of Method(typeof(A))
    }
}
````

Why it's needed:
1. Allows writing code generators that consume only syntax trees (no semantics) and dynamically determine operand type and what to do with it.
2. Syntax sugar.

Dart has this implemented and it comes very handy when writing code generators.

Update: As some have pointed out, the proposal may too much concentrate on the particular needs of OP. The suggestions is to focus on the syntax sugar part.

**Comment by svick** [comment]
I don't understand, why would a code generator be able to process `Method(A)`, but not `Method(typeof(A))`? Especially considering that without semantics, `A` is more ambiguous than `typeof(A)`, because `A` could reference a member or a local, while `typeof(A)` clearly references a type.

**Comment by edward-a** [comment]
> 
> 
> I don't understand, why would a code generator be able to process `Method(A)`, but not `Method(typeof(A))`? Especially considering that without semantics, `A` is more ambiguous than `typeof(A)`, because `A` could reference a member or a local, while `typeof(A)` clearly references a type.

Because generator doesn't have type info on this identifier since it's only dealing with syntax trees. Consider parsing an expression. Some identifiers will represent variables while others are class names. Having a way to pass all expression parameters into a function that dynamically decides what to do with those makes possible to write code generators that would rather obtain type info dynamically than deal with clumsy and hard-to-implement semantics analysis.

**Comment by CyrusNajmabadi** [comment]
> Because generator doesn't have type info

Why not update the generator to have type info?

**Comment by edward-a** [comment]
> 
> 
> > Because generator doesn't have type info
> 
> Why not update the generator to have type info?

Roslyn semantics analysis is very sluggish compared to syntax. There is a faster alternative which is building project separately with MSBuild and extracting type info through reflection wherever changes occur but this entails quite an amount of work and support.

**Comment by svick** [comment]
@edward-a If you want to, you can easily remove `typeof` from your code, for example by using `CSharpSyntaxRewriter`:

```c#
class RemoveTypeofVisitor : CSharpSyntaxRewriter
{
    public override SyntaxNode VisitTypeOfExpression(TypeOfExpressionSyntax node)
        => node.Type;
}

…

new RemoveTypeofVisitor().Visit(node)
```

The result of this will not be semantically correct, but I don't think that matters to you.

If you use this to preprocess code with `typeof`, your generator will only see the code in the form you like. I don't see why would this require changing the language.

**Comment by edward-a** [comment]
> 
> 
> @edward-a If you want to, you can easily remove `typeof` from your code, for example by using `CSharpSyntaxRewriter`:
> 
> ```cs
> class RemoveTypeofVisitor : CSharpSyntaxRewriter
> {
>     public override SyntaxNode VisitTypeOfExpression(TypeOfExpressionSyntax node)
>         => node.Type;
> }
> 
> …
> 
> new RemoveTypeofVisitor().Visit(node)
> ```
> 
> The result of this will not be semantically correct, but I don't think that matters to you.
> 
> If you use this to preprocess code with `typeof`, your generator will only see the code in the form you like. I don't see why would this require changing the language.

The source code is an expression and has to be compilabe. E.g. a = b + Math.Abs(c). To obtain type info dynamically in the generated code I need to be able to make a call like G(b, Math, Math.Abs, c). Second parameter is the problem.

**Comment by CyrusNajmabadi** [comment]
> Roslyn semantics analysis is very sluggish compared to syntax. 

That's surprising to me.  After all, semantic analysis needs to run in the sub-milliseconds range in order to support things like intellisense.

**Comment by CyrusNajmabadi** [comment]
> The source code is an expression and has to be compilabe.

That would be the case with: `Method(typeof(A))`.  It would be compilable.  You would then just trim out the `typeof` portion *yourself* when doing the source transformation.

> To obtain type info dynamically in the generated code I need to be able to make a call like G(b, Math, Math.Abs, c). Second parameter is the problem.

So have it be `G(b, typeof(Math), Math.Abs, c)`, then when doing your transformation, convert `typeof` as you desire.

**Comment by edward-a** [comment]
> 
> 
> > Roslyn semantics analysis is very sluggish compared to syntax.
> 
> That's surprising to me. After all, semantic analysis needs to run in the sub-milliseconds range in order to support things like intellisense.

It's super slow if you use from code.

**Comment by edward-a** [comment]
> 
> 
> > The source code is an expression and has to be compilabe.
> 
> That would be the case with: `Method(typeof(A))`. It would be compilable. You would then just trim out the `typeof` portion _yourself_ when doing the source transformation.
> 
> > To obtain type info dynamically in the generated code I need to be able to make a call like G(b, Math, Math.Abs, c). Second parameter is the problem.
> 
> So have it be `G(b, typeof(Math), Math.Abs, c)`, then when doing your transformation, convert `typeof` as you desire.

Code generator doesn't know which member to wrap with typeof because it doesn't have type info from the semantic analysis. It's only dealing with expressions like a = b + Math.Abs(c)

**Comment by svick** [comment]
@edward-a 

> The source code is an expression and has to be compilabe. E.g. a = b + Math.Abs(c). To obtain type info dynamically in the generated code I need to be able to make a call like G(b, Math, Math.Abs, c). Second parameter is the problem.

Isn't the third parameter also a problem? For this code to compile, `G` would have to have an overload that accepts `Func<int, int>` and I don't think there's a reasonable way of having that without semantic analysis.

**Comment by CyrusNajmabadi** [comment]
> It's super slow if you use from code.

I don't know what this means.  I use Roslyn "from code" all the time, and i can do tons of semantic operations per second without any issue.

**Comment by CyrusNajmabadi** [comment]
> Code generator doesn't know which member to wrap with typeof because it doesn't have type info from the semantic analysis. It's only dealing with expressions like a = b + Math.Abs(c)

i don't see how this could work effectively for so many cases in C#.  It's not just `types` that cannot be passed effectively do `object`.  Lots of other constructs don't have implicit conversions either.  Can you clarify how you would expect those to work?

**Comment by CyrusNajmabadi** [comment]
> because it doesn't have type info from the semantic analysis.

I still need a good explanation of why using Roslyn to get this type information is out of scope.  

I just checked and i was able to do *1 million* semantic model operations in roslyn in just over 2 seconds.  This was on a file like:

```c#
using System;
class C {
    void Foo(int a, int b) {
        var x = a + Math.Abs(b);
    }
    static void Main() { }
}
```

**Comment by edward-a** [comment]
> 
> 
> @edward-a
> 
> > The source code is an expression and has to be compilabe. E.g. a = b + Math.Abs(c). To obtain type info dynamically in the generated code I need to be able to make a call like G(b, Math, Math.Abs, c). Second parameter is the problem.
> 
> Isn't the third parameter also a problem? For this code to compile, `G` would have to have an overload that accepts `Func<int, int>` and I don't think there's a reasonable way of having that without semantic analysis.

Yes, it's a also a problem. But this problem is solvable since we only care about function return type to be able to parse member accesses like A.B(C).D.E. And since code generator already knows from the syntax it's a method call, we can use the code below to wrap the function and obtain it's return type dynamically:

````csharp
    class FuncWrapper<T>
    {
      public static FuncWrapper<T> Create(Func<T> f) {
        return new FuncWrapper<T>();
      }
    }

    class FuncWrapperFactory
    {
      public static FuncWrapper<T> Create<T>(Func<T> f) {
        return new FuncWrapper<T>();
      }
    }  
    
    G(b, Math, FuncWrapperFactory.Create(() => System.Math.Abs(c)), c)
````

**Comment by CyrusNajmabadi** [comment]
> since we only care

Here's the issue.  You are proposing an extremely limited change to the language to be sufficient for "only" your "care".  The language can't really be designed in this fashion.  What happens when the next person comes along and says "but i need untyped functions of arbitrary complexity to be supported"? 

> `FuncWrapperFactory.Create(() => System.Math.Abs(c), c)`

I don't see how this compiles.  You don't have an overload of `Create` that takes two args.

**Comment by edward-a** [comment]
> 
> 
> > It's super slow if you use from code.
> 
> I don't know what this means. I use Roslyn "from code" all the time, and i can do tons of semantic operations per second without any issue.



> 
> 
> > since we only care
> 
> Here's the issue. You are proposing an extremely limited change to the language to be sufficient for "only" your "care". The language can't really be designed in this fashion. What happens when the next person comes along and says "but i need untyped functions of arbitrary complexity to be supported"?

It's up to, i am not pushing anyone. You manage the language and it's up to you to decide whether it's appropriate or not to support what i have proposed. Dart supports it and allows me to build nice code generators. I myself come from C# background and was just curious how hard it would be to have the same capability. Again, not pushing anyone.

> 
> > `FuncWrapperFactory.Create(() => System.Math.Abs(c), c)`
> 
> I don't see how this compiles. You don't have an overload of `Create` that takes two args.

That's a typo, fixed.

**Comment by edward-a** [comment]
> 
> 
> > It's super slow if you use from code.
> 
> I don't know what this means. I use Roslyn "from code" all the time, and i can do tons of semantic operations per second without any issue.

To generate semantics you need to use MSBuild to load the entire solution the time you generate code. It's pretty slow to do from the code. The solution would be to run a separate MSBuild server to build code for the analysis and extract type info through reflection.

**Comment by CyrusNajmabadi** [comment]
>It's up to, i am not pushing anyone. 

Sure.  My point is: language changes will only happen with *super compelling cases*.  And that means making really good cases for them, and not having them be extremely niche.   :)

**Comment by edward-a** [comment]
> 
> 
> > because it doesn't have type info from the semantic analysis.
> 
> I still need a good explanation of why using Roslyn to get this type information is out of scope.
> 
> I just checked and i was able to do _1 million_ semantic model operations in roslyn in just over 2 seconds. This was on a file like:
> 
> ```cs
> using System;
> class C {
>     void Foo(int a, int b) {
>         var x = a + Math.Abs(b);
>     }
>     static void Main() { }
> }
> ```

When using a real life solution with multiple dependencies the waiting time to obtain semantics is quite long.

**Comment by CyrusNajmabadi** [comment]
> To generate semantics you need to use MSBuild to load the entire solution 

No you don't.  You could, for example, just have some source files and metadata references.   :)

But say you want to use msbuild.  Ok, so you use it... why is that a problem?  

> The solution would be to run a separate MSBuild server to build code for the analysis and extract type info through reflection.

Ok.  Seems like you have a good solution here that is orders of magnitude cheaper than changing the lanuage to support this niche case... And now, with semantics, you don't have to be concerned with all these sorts of strange cases.  You can just figure everything out tht you need, doing whatever sort of code-gen you want.   :)

**Comment by edward-a** [comment]
> 
> 
> > It's up to, i am not pushing anyone.
> 
> Sure. My point is: language changes will only happen with _super compelling cases_. And that means making really good cases for them, and not having them be extremely niche.

I don't think support for code generators based on syntax trees is niche but again you should know better.

**Comment by CyrusNajmabadi** [comment]
> When using a real life solution with multiple dependencies the waiting time to obtain semantics is quite long.

Define 'quite long'.  Can you also explain your requirements where that isn't an acceptable cost to pay?

**Comment by CyrusNajmabadi** [comment]
> I don't think support for code generators based on syntax trees is niche but again you should know better.

You're the first person to ask for this particular feature.  There are thousands of requests to roslyn over its lifetime... This is def niche :)

**Comment by CyrusNajmabadi** [comment]
You may want to consider https://github.com/dotnet/csharplang/issues/107 and check if that would provide a solution for you.  It would likely give you syntax transformations with practically free semantic-ops (since it would run in the context of the compiler, which will have already computed the semantic info).

**Comment by edward-a** [comment]
> 
> 
> > I don't think support for code generators based on syntax trees is niche but again you should know better.
> 
> You're the first person to ask for this particular feature. There are thousands of requests to roslyn over its lifetime... This is def niche :)

I don't think the benefits it brings are niche. Code generator are quite popular now.

**Comment by edward-a** [comment]
> 
> 
> > When using a real life solution with multiple dependencies the waiting time to obtain semantics is quite long.
> 
> Define 'quite long'. Can you also explain your requirements where that isn't an acceptable cost to pay?

Seconds

**Comment by CyrusNajmabadi** [comment]
> Seconds

Why are seconds not acceptable in this domain?  

> I don't think the benefits it brings are niche. Code generator are quite popular now.

Sure.  but roslyn supports this scenario.  You have chosen some specific constraints taht don't allow you to use Roslyn though.  For example, i have written many code generators (for several companies).  Using semantics has not been a problem.

**Comment by edward-a** [comment]
> 
> 
> > To generate semantics you need to use MSBuild to load the entire solution
> 
> No you don't. You could, for example, just have some source files and metadata references. :) 
Elaborate please. Say, i have a .cs file for which i need to generate code. I know what project it is from, so i use MSBuild to load it and be able to build semantic model. You know a different approach to this? Please share.

> But say you want to use msbuild. Ok, so you use it... why is that a problem?
As i said, it's sluggish when used from code.

> > The solution would be to run a separate MSBuild server to build code for the analysis and extract type info through reflection.
> 
> Ok. Seems like you have a good solution here that is orders of magnitude cheaper than changing the lanuage to support this niche case... And now, with semantics, you don't have to be concerned with all these sorts of strange cases. You can just figure everything out tht you need, doing whatever sort of code-gen you want. :)
No, the solution is not good since it's a separate server like Omnisharp and it will require an effort to develop and maintain it. I'd rather drop the idea and not port my library from Dart to C# and share it with the community.

**Comment by CyrusNajmabadi** [comment]
> No, the solution is not good since it's a separate server like Omnisharp and it will require an effort to develop and maintain it. I'd rather drop the idea and not port my library from Dart to C# and share it with the community.

But having the language change to support these constructs for your single case would certainly cost more, right?  Probably by orders of magnitude.

>  I'd rather drop the idea and not port my library from Dart to C# and share it with the community.

Can you link me to the library?  I can see about helping out there.  It feels like there would be solutions that could be used to make it work out that could still be fast (including without multiple seconds of MSBuild).

**Comment by edward-a** [comment]
> 
> 
> > Seconds
> 
> Why are seconds not acceptable in this domain?
> 
> > I don't think the benefits it brings are niche. Code generator are quite popular now.
> 
> Sure. but roslyn supports this scenario. You have chosen some specific constraints taht don't allow you to use Roslyn though. For example, i have written many code generators (for several companies). Using semantics has not been a problem.

Did your generators generate code once you save the file? Having seconds to wait is not acceptable in this scenario.

**Comment by CyrusNajmabadi** [comment]
> Did your generators generate code once you save the file?

Yes. 

>  Having seconds to wait is not acceptable in this scenario.

For 99% of my projects it's been ok.  For one company i wrote a little daemon to just stay resident and process inputs.  I think it took me 15 min or less to just write that...  it never seemed to be an issue.

**Comment by edward-a** [comment]
> 
> 
> > No, the solution is not good since it's a separate server like Omnisharp and it will require an effort to develop and maintain it. I'd rather drop the idea and not port my library from Dart to C# and share it with the community.
> 
> But having the language change to support these constructs for your single case would certainly cost more, right? Probably by orders of magnitude.

I don't think it's my single case. I might get widely adopted once you introduce it. And it's not much of a change to the compiler. Just a syntax sugar, passing type name as a parameter instead of typeof(class name).
> 
> > I'd rather drop the idea and not port my library from Dart to C# and share it with the community.
> 
> Can you link me to the library? I can see about helping out there. It feels like there would be solutions that could be used to make it work out that could still be fast (including without multiple seconds of MSBuild).

It's not publicly available yet. I was playing with an idea to publish it in both Dart and C# at the same time (i am a fan of C#). But, obviously, even if you accept the change, it will take a while before it gets implemented. I am still looking into how to make it work in C# and if i don't come up with anything useful i will release in Dart and provide you with a link.

**Comment by edward-a** [comment]
> 
> 
> > Did your generators generate code once you save the file?
> 
> Yes.
> 
> > Having seconds to wait is not acceptable in this scenario.
> 
> For 99% of my projects it's been ok. For one company i wrote a little daemon to just stay resident and process inputs. I think it took me 15 min or less to just write that... it never seemed to be an issue.

I was planning to release the library as a nuget package. I don't think nuget supports installing standalone daemons/services.

**Comment by CyrusNajmabadi** [comment]
> I don't think it's my single case. 

Can you clarify why you think that?  Many projects do source generation/transformation using Roslyn.  But you're the first i've seen that has had a hard requirement of not being able to know the semantics of what is being transformed.

>  Just a syntax sugar, passing type name as a parameter instead of typeof(class name).

Any language change is massively expensive (it would require dozens of people to spend hundreds to thousands of hours to actually spec this all out, doc it, implement it, test it, disseminate it, etc.)  To be accepted, the feature has to demonstrate huge value that warrants the effort that needs to go into it.  

Furthermore, it needs to be broadly applicable.  So it can't just solve this exact case for a single customer.  After all, what happens when the next person comes along and says "oh, those cases that @edward-a didn't care about... well i care about them.  you did this for him, you need to do this for me."

**Comment by CyrusNajmabadi** [comment]
> I was planning to release the library as a nuget package. I don't think nuget supports installing standalone daemons/services.

You can make the daemon available however you want.  If your customers say that the second or so of msbuild time is not acceptable, you can guide them how to launch your daemon and have that speed things up.

**Comment by CyrusNajmabadi** [comment]
Note: you might want to consdier going to gitter.im/dotnet/roslyn or to the C# discord channel at this point.  It sounds like it would be more conducive for brainstorming potential solutions to what you're trying to do.  It would also result in solutions that likely would be feasible within minutes or hours (unlike a language change, while will likely take years).

**Comment by DavidArno** [comment]
The code example in the OP is a little odd and I don't follow the logic of code generators benefiting from this, but for a method,
```cs
void Foo(Type x) { ...
```
being able to call it via `Foo(SomeType)` rather than `Foo(typeof(SomeType))` definitely feels more elegant and removes some pointless "noise", so I'm generally in favour of this proposal.

**Comment by FiniteReality** [comment]
> The code example in the OP is a little odd and I don't follow the logic of code generators benefiting from this, but for a method,
> 
> ```cs
> void Foo(Type x) { ...
> ```
> 
> being able to call it via `Foo(SomeType)` rather than `Foo(typeof(SomeType))` definitely feels more elegant and removes some pointless "noise", so I'm generally in favour of this proposal.

I'd generally agree, but it would be a source breaking change. For example:

```cs
static class CallerHelpers
{
    internal static void Call(Type callerType) { throw null; }
    internal static void Call(Caller caller) { throw null; }
}

class CallerInfo
{
    public Caller Caller { get; set; }

    public void DoSomething()
    {
        CallerHelpers.Call(Caller);
    }
}
```

In this case, while contrived, CallerInfo.DoSomething can *only* call `CallerHelpers.Call(Caller caller)` currently. If we allowed passing type names as a parameter without `typeof`, this would break the code as it is now ambiguous between `CallerHelpers.Call(this.Caller)` and `CallerHelpers.Call(typeof(Caller))`.

I've seen code not entirely unlike this in multiple codebases.

**Comment by DavidArno** [comment]
@FiniteReality, ah good point re breaking change conflicts like that. Oh well, back to the drawing board 😄

**Comment by edward-a** [comment]
> 
> 
> > The code example in the OP is a little odd and I don't follow the logic of code generators benefiting from this, but for a method,
> > ```cs
> > void Foo(Type x) { ...
> > ```
> > 
> > 
> > being able to call it via `Foo(SomeType)` rather than `Foo(typeof(SomeType))` definitely feels more elegant and removes some pointless "noise", so I'm generally in favour of this proposal.
> 
> I'd generally agree, but it would be a source breaking change. For example:
> 
> ```cs
> static class CallerHelpers
> {
>     internal static void Call(Type callerType) { throw null; }
>     internal static void Call(Caller caller) { throw null; }
> }
> 
> class CallerInfo
> {
>     public Caller Caller { get; set; }
> 
>     public void DoSomething()
>     {
>         CallerHelpers.Call(Caller);
>     }
> }
> ```
> 
> In this case, while contrived, CallerInfo.DoSomething can _only_ call `CallerHelpers.Call(Caller caller)` currently. If we allowed passing type names as a parameter without `typeof`, this would break the code as it is now ambiguous between `CallerHelpers.Call(this.Caller)` and `CallerHelpers.Call(typeof(Caller))`.
> 
> I've seen code not entirely unlike this in multiple codebases.

I believe this gets resolved easily by making compiler choose the second method in such cases. Or am i missing something?

**Comment by HaloFour** [comment]
It complicates the implementation.  For a proposal that basically benefits a single person writing a very niche tool that intentionally skips using the built-in components for syntax analysis it's already exceptionally unlikely to be considered.  Upping the difficulty, by any degree, only further ensures this.

**Comment by edward-a** [comment]
> 
> that intentionally skips using the built-in components for syntax analysis.

That is just not true. The syntax analysis components are fully utilized. But anyway, i suggest skipping on the OP's particular needs and review the syntax sugar benefits and complications. Updated the OP's post appropriately.

**Comment by HaloFour** [comment]
Eliminating 8 characters around a relatively uncommon operation doesn't seem like a particularly big benefit, especially since it involves the compiler having to interpret the token as potentially two different types of identifiers.  Even eliminating the whole syntax analysis and source generator use case as described in the OP there's so little benefit here.

**Comment by FiniteReality** [comment]
In most cases in generators, you know whether something is going to be used as a type or a variable anyway, since you're the one *generating* the code. I can't think of a case where this isn't the case, and even in generators which aren't based on Roslyn you could probably use some form of tagging to indicate whether a value is a type or not.

**Comment by edward-a** [comment]
> 
> 
> In most cases in generators, you know whether something is going to be used as a type or a variable anyway, since you're the one _generating_ the code.

The generated code is based on expressions written in C#, so without semantics there is no way to statically resolve whether a type or variable is used for member accessing. It should be possible to resolve dynamically if the proposal goes through.

**Comment by Joe4evr** [comment]
> I believe this gets resolved easily by making compiler choose the second method in such cases. Or am i missing something?

That would be a big change in behavior.
> The generated code is based on expressions written in C#

Can you give an example expression and how your generator interprets that to cause this ambiguity?

**Comment by edward-a** [comment]
> 
> 
> > I believe this gets resolved easily by making compiler choose the second method in such cases. Or am i missing something?
> 
> That would be a big change in behavior.

Can you elaborate on this? It doesn't seem that it changes the current behavior, although i might be missing something here.
> 
> > The generated code is based on expressions written in C#
> 
> Can you give an example expression and how your generator interprets that to cause this ambiguity?

The simplest example would be accessing static class member, e.g. Math.Abs(). There is no way to know if Math is a type or variable from syntax analysis. You could however figure this out if you were able to pass Math as an argument to a function and use reflection.

**Comment by Joe4evr** [comment]
> Can you elaborate on this? It doesn't seem that it changes the current behavior, although i might be missing something here.

As stated, code currently can only pick the first overload. However, with the change you've suggested, that same code will suddenly pick the second overload on recompilation, which is s a change in behavior and violates the [Principle of Least Surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

**Comment by edward-a** [comment]
> 
> 
> > Can you elaborate on this? It doesn't seem that it changes the current behavior, although i might be missing something here.
> 
> As stated, code currently can only pick the first overload. However, with the change you've suggested, that same code will suddenly pick the second overload on recompilation, which is s a change in behavior and violates the [Principle of Least Surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

I don't see it picking up the first overload, it's definitively the second overload that gets called in DoSomething.

````csharp
static class CallerHelpers
{
    internal static void Call(Type callerType) { throw null; }
    internal static void Call(Caller caller) { throw null; }
}

class CallerInfo
{
    public Caller Caller { get; set; }

    public void DoSomething()
    {
        CallerHelpers.Call(Caller);
    }
}
````
Implementing the proposal and making compiler favor the second call over first doesn't seem to change current behavior.

**Comment by Joe4evr** [comment]
Oh, that's my bad; I thought `Call(Type)` was the second overload (which made a lot of sense in my head since a reflection-based overload should be a lower priority than a non-reflection one).

**Comment by SteffenBlake** [comment]
I would enjoy this as well, many times I have methods that take in a string as a "nameof" that they _want_ to receive, but I have no way to enforce that.

I would simply expect the syntax to be a variable that is coalesced to a stringifiable variable via nameof, automatically, but still retains the type safety if needed. IE I may still want to specify "I will want the name of this variable, but that variable still needs to be an int!"

```
public void LogIntName(name<int> n) {
    // n will be a 'nameof'<int> here, which can be handled as a string.
    Console.WriteLine(n);
}

var foo = 5;
var bar = 6;

LogIntName(foo); // Writes "foo" to the console
LogIntName(bar); // writes "bar" to the console=
```

Something like this.

**Comment by yaakov-h** [comment]
@SteffenBlake that sounds more like #287 to me.
